File: react-native-nitro-player/android/src/main/java/com/margelo/nitro/nitroplayer/HybridTrackPlayer.kt
package com.margelo.nitro.nitroplayer

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.margelo.nitro.nitroplayer.core.TrackPlayerCore

@DoNotStrip
@Keep
class HybridTrackPlayer : HybridTrackPlayerSpec() {
    private val core: TrackPlayerCore

    init {
        val context =
            NitroModules.applicationContext ?: throw IllegalStateException("React Context is not initialized")
        core = TrackPlayerCore.getInstance(context)
    }

    @DoNotStrip
    @Keep
    override fun play() {
        core.play()
    }

    @DoNotStrip
    @Keep
    override fun pause() {
        core.pause()
    }

    override fun playSong(
        songId: String,
        fromPlaylist: String?,
    ): Promise<Unit> =
        Promise.async {
            core.playSong(songId, fromPlaylist)
        }

    @DoNotStrip
    @Keep
    override fun skipToNext() {
        core.skipToNext()
    }

    @DoNotStrip
    @Keep
    override fun skipToPrevious() {
        core.skipToPrevious()
    }

    @DoNotStrip
    @Keep
    override fun seek(position: Double) {
        core.seek(position)
    }

    override fun addToUpNext(trackId: String): Promise<Unit> =
        Promise.async {
            core.addToUpNext(trackId)
        }

    override fun playNext(trackId: String): Promise<Unit> =
        Promise.async {
            core.playNext(trackId)
        }

    override fun getActualQueue(): Promise<Array<TrackItem>> =
        Promise.async {
            core.getActualQueue().toTypedArray()
        }

    override fun getState(): Promise<PlayerState> =
        Promise.async {
            core.getState()
        }

    @DoNotStrip
    @Keep
    override fun setRepeatMode(mode: RepeatMode): Boolean = core.setRepeatMode(mode)

    @DoNotStrip
    @Keep
    override fun getRepeatMode(): RepeatMode = core.getRepeatMode()

    override fun onChangeTrack(callback: (track: TrackItem, reason: Reason?) -> Unit) {
        core.addOnChangeTrackListener(callback)
    }

    override fun onPlaybackStateChange(callback: (state: TrackPlayerState, reason: Reason?) -> Unit) {
        core.addOnPlaybackStateChangeListener(callback)
    }

    override fun onSeek(callback: (position: Double, totalDuration: Double) -> Unit) {
        core.addOnSeekListener(callback)
    }

    override fun onPlaybackProgressChange(callback: (position: Double, totalDuration: Double, isManuallySeeked: Boolean?) -> Unit) {
        core.addOnPlaybackProgressChangeListener(callback)
    }

    @DoNotStrip
    @Keep
    override fun configure(config: PlayerConfig) {
        core.configure(
            androidAutoEnabled = config.androidAutoEnabled,
            carPlayEnabled = config.carPlayEnabled,
            showInNotification = config.showInNotification,
            lookaheadCount = config.lookaheadCount?.toInt(),
        )
    }

    @Keep
    override fun onAndroidAutoConnectionChange(callback: (Boolean) -> Unit) {
        core.onAndroidAutoConnectionChange = callback
    }

    @Keep
    override fun isAndroidAutoConnected(): Boolean = core.isAndroidAutoConnected()

    @DoNotStrip
    @Keep
    override fun setVolume(volume: Double): Boolean = core.setVolume(volume)

    @DoNotStrip
    @Keep
    override fun skipToIndex(index: Double): Promise<Boolean> =
        Promise.async {
            core.skipToIndex(index.toInt())
        }

    override fun updateTracks(tracks: Array<TrackItem>): Promise<Unit> =
        Promise.async {
            core.updateTracks(tracks.toList())
        }

    override fun getTracksById(trackIds: Array<String>): Promise<Array<TrackItem>> =
        Promise.async {
            core.getTracksById(trackIds.toList()).toTypedArray()
        }

    override fun getTracksNeedingUrls(): Promise<Array<TrackItem>> =
        Promise.async {
            core.getTracksNeedingUrls().toTypedArray()
        }

    override fun getNextTracks(count: Double): Promise<Array<TrackItem>> =
        Promise.async {
            core.getNextTracks(count.toInt()).toTypedArray()
        }

    override fun getCurrentTrackIndex(): Promise<Double> =
        Promise.async {
            core.getCurrentTrackIndex().toDouble()
        }

    override fun setPlaybackSpeed(speed: Double): Promise<Unit> =
        Promise.async {
            core.setPlayBackSpeed(speed)
            Unit
        }

    override fun getPlaybackSpeed(): Promise<Double> =
        Promise.async {
            core.getPlayBackSpeed()
        }

    override fun onTracksNeedUpdate(callback: (tracks: Array<TrackItem>, lookahead: Double) -> Unit) {
        core.addOnTracksNeedUpdateListener { tracks, lookahead ->
            callback(tracks.toTypedArray(), lookahead.toDouble())
        }
    }
}

File: react-native-nitro-player/ios/HybridTrackPlayer.swift
//
//  HybridTrackPlayer.swift
//  NitroPlayer
//
//  Created by Ritesh Shukla on 10/12/25.
//

import Foundation
import NitroModules

/// Hybrid implementation of TrackPlayerSpec for iOS
/// Bridges Nitro modules with the native TrackPlayerCore implementation
final class HybridTrackPlayer: HybridTrackPlayerSpec {
  // MARK: - Properties

  private let core: TrackPlayerCore

  // MARK: - Initialization

  override init() {
    core = TrackPlayerCore.shared
    super.init()
  }

  // MARK: - Playback Control

  func play() throws {
    core.play()
  }

  func pause() throws {
    core.pause()
  }

  func playSong(songId: String, fromPlaylist: String?) throws -> Promise<Void> {
    return Promise.async {
      self.core.playSong(songId: songId, fromPlaylist: fromPlaylist)
    }
  }

  func skipToNext() throws {
    core.skipToNext()
  }

  func skipToPrevious() throws {
    core.skipToPrevious()
  }

  func seek(position: Double) throws {
    core.seek(position: position)
  }

  func addToUpNext(trackId: String) throws -> Promise<Void> {
    return Promise.async {
      self.core.addToUpNext(trackId: trackId)
    }
  }

  func playNext(trackId: String) throws -> Promise<Void> {
    return Promise.async {
      self.core.playNext(trackId: trackId)
    }
  }

  func getActualQueue() throws -> Promise<[TrackItem]> {
    return Promise.async {
      return self.core.getActualQueue()
    }
  }

  func getState() throws -> Promise<PlayerState> {
    return Promise.async {
      return self.core.getState()
    }
  }

  func setRepeatMode(mode: RepeatMode) throws -> Bool {
    return core.setRepeatMode(mode: mode)
  }

  // MARK: - Configuration

  func configure(config: PlayerConfig) throws {
    core.configure(
      androidAutoEnabled: config.androidAutoEnabled,
      carPlayEnabled: config.carPlayEnabled,
      showInNotification: config.showInNotification
    )
  }

  // MARK: - Event Callbacks

  func onChangeTrack(callback: @escaping (TrackItem, Reason?) -> Void) throws {
    NitroPlayerLogger.log("HybridTrackPlayer", "onChangeTrack callback registered")
    core.addOnChangeTrackListener(owner: self, callback)
  }

  func onPlaybackStateChange(callback: @escaping (TrackPlayerState, Reason?) -> Void) throws {
    NitroPlayerLogger.log("HybridTrackPlayer", "onPlaybackStateChange callback registered")
    core.addOnPlaybackStateChangeListener(owner: self, callback)
  }

  func onSeek(callback: @escaping (Double, Double) -> Void) throws {
    NitroPlayerLogger.log("HybridTrackPlayer", "onSeek callback registered")
    core.addOnSeekListener(owner: self, callback)
  }

  func onPlaybackProgressChange(callback: @escaping (Double, Double, Bool?) -> Void) throws {
    NitroPlayerLogger.log("HybridTrackPlayer", "onPlaybackProgressChange callback registered")
    core.addOnPlaybackProgressChangeListener(owner: self, callback)
  }

  // MARK: - Android Auto (iOS No-op)

  /// iOS doesn't support Android Auto, so this is a no-op
  /// - Parameter callback: Callback that will never be invoked
  func onAndroidAutoConnectionChange(callback: @escaping (Bool) -> Void) throws {
    // iOS doesn't have Android Auto, so this is a no-op
  }

  /// iOS doesn't support Android Auto, always returns false
  /// - Returns: Always returns false on iOS
  func isAndroidAutoConnected() throws -> Bool {
    return false
  }

  func skipToIndex(index: Double) throws -> Promise<Bool> {
    return Promise.async {
      return self.core.skipToIndex(index: Int(index))
    }
  }

  // MARK: - Volume Control

  func setVolume(volume: Double) throws -> Bool {
    return core.setVolume(volume: volume)
  }
}
